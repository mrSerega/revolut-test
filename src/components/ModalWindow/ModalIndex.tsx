import React from 'react';
import { ModalKind } from '../../typings/modals';
import { ModalWindow } from './ModalWindow';

export interface ModalIndexStateProps {
    modalKind: ModalKind | null
    message?: string
}

export interface ModalIndexDispatchProps {
    onClose: () => void;
}

export type ModalIndexProps =
    & ModalIndexStateProps
    & ModalIndexDispatchProps

export class ModalIndex extends React.Component<ModalIndexProps> {
    readonly = 'modal-index'

    render() {
        const {
            modalKind,
            onClose,
            message
        } = this.props

        if (!modalKind) {
            return null
        }

        switch (modalKind) {
            case ModalKind.CommonModal:
                return <ModalWindow
                    onClose={onClose}
                    message="CommonModal"
                />
            case ModalKind.ErrorModal:
                return <ModalWindow
                    onClose={onClose}
                    message={message || ''}
                />
            default:
                const unknownModalKind: never = modalKind
                throw new Error(`Unknown modal kind: ${unknownModalKind}`)
        }
    }
}