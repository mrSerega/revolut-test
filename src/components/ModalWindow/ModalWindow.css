.modal-window__cover{
    background-color: rgba(255,255,255,0.3);
    position: absolute;
    left: 0;
    top: 0;
    width: 100%;
    height: 100%;
    z-index: 100;
}

.modal-window__error-dialog {
    background-color: white;
    margin: 40px;
    height: calc(100% - 104px);
    width: calc(100% - 104px);
    border-radius: 8px;
    padding: 12px;
    position: relative;
}

.modal-window__error-title {
    font-size: 40px;
    font-weight: 800;
}

.modal-window__error-content {
    margin-top: 32px;
    font-size: 20px;
    line-height: 24px;
}

.modal-window__error-footer {
    position: absolute;
    bottom: 12px;
}

.modal-window__button {
    position: absolute;
    bottom: 0;
    font-size: 20px;
    padding: 8px;
    border: 1px solid black;
    border-radius: 8px;
    cursor: pointer;
}