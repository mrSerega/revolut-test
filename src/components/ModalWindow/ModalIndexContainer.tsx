import { connect } from 'react-redux';
import { RootState } from '../../states/indexState';
import { Dispatch } from 'redux';
import { ModalIndex, ModalIndexDispatchProps, ModalIndexStateProps } from './ModalIndex';
import { ModalSelector } from '../../states/modalState';
import { toggleModal } from '../../actions/modalActions';

const mapStateToProps = (state: RootState): ModalIndexStateProps => {
    const modalKind = ModalSelector.getCurrentModalKind(state)
    const message = ModalSelector.getModalMessage(state)

    return {
        modalKind,
        message
    }
}

const mapDispatchToProps = (dispatch: Dispatch): ModalIndexDispatchProps => ({
    onClose: () => dispatch(toggleModal({modalKind: undefined}))
})

export const ModalIndexContainer = connect<
    ModalIndexStateProps,
    ModalIndexDispatchProps,
    {},
    RootState
>(
    mapStateToProps,
    mapDispatchToProps
)(ModalIndex)