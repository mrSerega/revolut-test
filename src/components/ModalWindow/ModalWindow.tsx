import React from 'react';

import './ModalWindow.css'

export interface ModalWindowOwnProps {
    onClose: () => void;
    message: string
}

export type ModalWindowProps =
    & ModalWindowOwnProps

export class ModalWindow extends React.Component<ModalWindowProps> {
    readonly blockName = 'modal-window'

    componentDidMount() {
        window.addEventListener('keydown', this.focusLockListener);
    }

    componentWillUnmount() {
        window.removeEventListener('keydown', this.focusLockListener)
    }


    render(){

        const {
            onClose,
            message
        } = this.props

        return <div className={this.blockName + '__cover'}>
            <div className={this.blockName + '__error-dialog'}>
                <div className={this.blockName + '__error-title'}>
                    Oops!
                </div>
                <div className={this.blockName + '__error-content'}>
                    {message}
                </div>
                <div className={this.blockName + '__error-footer'}>
                    <div
                        className={
                            this.blockName + '__button'
                        }
                        onClick={onClose}
                    >
                        OK
                    </div>
                </div>
            </div>
        </div>
    }

    private focusLockListener = (evt: KeyboardEvent) => {
        if (evt.key === 'Tab') {
            evt.preventDefault();
        }
    }
}