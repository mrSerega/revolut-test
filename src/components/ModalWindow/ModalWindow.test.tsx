import React from 'react';
import renderer from 'react-test-renderer';
import { ModalWindow } from './ModalWindow';

test('ModalWindow', () => {
    const cmp = renderer.create(
        <ModalWindow
            onClose={() => {}}
            message="message"
        />
    )
    const tree = cmp.toJSON()
    expect(tree).toMatchSnapshot()
})