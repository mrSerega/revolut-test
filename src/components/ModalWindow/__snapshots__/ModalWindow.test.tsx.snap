// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalWindow 1`] = `
<div
  className="modal-window__cover"
>
  <div
    className="modal-window__error-dialog"
  >
    <div
      className="modal-window__error-title"
    >
      Oops!
    </div>
    <div
      className="modal-window__error-content"
    >
      message
    </div>
    <div
      className="modal-window__error-footer"
    >
      <div
        className="modal-window__button "
        onClick={[Function]}
      >
        OK
      </div>
    </div>
  </div>
</div>
`;
